effekt/shared/src/main/scala/effekt/core/Recursive.scala

```diff
@@ -0,0 +1,132 @@
+package effekt.core
+
+import scala.collection.mutable
+
+// TODO: Is there a simpler way of iterating the functions and calls (like a fold)?
+//       Reachable etc. use similar recursions
+//       We could probably remove a lot of recursion here!
+
+/**
+ * Gather all functions, their arguments, and arguments of their recursive calls
+ */
+
+ case class FunctionGathering(
+    cparams: List[Id],
+    vparams: List[Id],
+    bparams: List[Id],
+    cargs: mutable.Set[List[Block]],
+    vargs: mutable.Set[List[Pure]],
+    bargs: mutable.Set[List[Block]]
+)
+
+class Recursive(
+  val defs: mutable.Map[Id, FunctionGathering],
+  var stack: List[Id]
+) {
+  def process(d: Definition): Unit =
```

Comment:
I am not sure if this applicable or sensible here, but have you tried to use [`Structural` macro](https://github.com/effekt-lang/effekt/blob/042dc2d2691e54c76502ec104a50f6c053e5bbcc/effekt/shared/src/main/scala/effekt/core/Tree.scala#L557) for automatically taking care of all trivial structural traversals you perform here?

Reply:
I think it is fine as explicit recursion for now, this way we can discuss cases ;)